test/BatchProcessor.test.ts

```diff
@@ -26,15 +27,277 @@ describe('BatchProcessor', () => {
+          }
+        }
+      }
+    })
+    await seneca.ready()
+    
+    const process = seneca.export('BatchProcessor/process')
+    
+    let state = { c: 0, msgs: [] }
+    let out, ctx, report
+    
+    seneca.message('aim:bar', async function(msg) {
+      state.ok = true
+      state.c++
+      // console.log('MSG: ', msg)
+      state.msgs.push(seneca.util.clean(msg))
+      return { ok: true, now: Date.now() }
+    })
+    
+    seneca.message('aim:foo,color:red', async function(msg) {
+    
+      out = { ok: true, planet: 'mars' }
+      ctx = { place: { order: 1 } }
+    
+      out = await process(this, ctx, out)
+    
+      await wait(111)
+    
+      // console.log(out, state, ctx)
+    
+      expect(state).toEqual({
+        ok: true,
+        c: 2,
+        msgs: [
+          { aim: 'bar', color: 'blue', planet: 'mars', order: 1 },
+          { aim: 'bar', color: 'green', planet: 'mars', order: 1 }
+        ]
+      })
+      expect(ctx.result$.length).toEqual(2)
+    
+    })
+    
+    await seneca.post('aim:foo,color:red')
+  
+  })
+  
+  describe('more complex', () => {
+    const seneca = makeSeneca({
+      send: {
+        mode: 'async', // wait for transition, global setting
+      },
+      where: {
+        'aim:foo,color:red': {
+          match: {
+            '*': { // catch all if no other patterns match
+              // Create BatchMonitor entry if ctx.BatchMonitorEntry$ defined 
+              entry: 'fail' // entry state, entry.info={why:'batch-process-no-match'}
+            },
+            'ok:false': {
+              entry: {
+                state: 'fail',
+                info: {
+                  why: 'out~why'
+                }
+              },
+              send: { // if single msg, no array needed
+                // ctx has original message in msg$
+                // out~ means entire contents of out object
+                msg: 'aim:monitor,fail:msg,msg:ctx~msg$,out:out~'
+              }
+            },
+            'ok:true': { // matches are in same Patrun set, so usual Seneca pattern rules apply
+              entry: 'done', // only created after all msgs sent
+              send: [ // zero or more next messages
+                {
+                  msg: {
+                    aim: 'bar',
+                    color: 'blue',
+                    planet: 'out~planet', // dot path ref
+                    order: 'ctx~place.order~Number' // Gubu validation expression
+                  }
+                },
+                {
+                  mode: 'sync', // use .act, don't await
+                  msg: 'aim:bar,color:green,planet:out~planet',
+                  body: { // msg has precedence
+                    order: 'ctx~place.order~Number'
+                  }
+                }
+              ]
+            }
+          }
+        }
+      }
+    })
+    
+    let state = { c: 0 }
+    let out, ctx, report, bme
+    
+    let process, batch
+    
+    beforeAll(async () => {
```

Comment:
this seems to be repeatedly adding the same messages to the same instance
let's not use beforeAll in any case as it relies on a global var
instead move this code to a test utility function makeSeneca and call that inside each test - this guarantees separate seneca instances for each test